### ISSUE TITLE
Validating empty array against spec should show better error

### ISSUE THREAD

**Issue by rly** [hdmf-dev/hdmf#317]
If an attribute spec specifies a dtype text with or without a shape, and an empty numpy array is validated against that spec, the error returned is: `IndexError: index 0 is out of bounds for axis 0 with size 0`. This is not very helpful.

Is an empty array allowed for `shape: None` (1D array with any length) ?

To clarify, what should be the correct validation behavior for the following cases:
1) Spec specifies `shape: None` and `dtype: float` and test value is empty array with type float -- ShapeError?
2) Spec specifies `shape: None` and `dtype: text` and test value is empty array with type float -- ShapeError and DtypeError?
3) Spec specifies `shape: None` and `dtype: text` and test value is empty list -- ShapeError?
4) Spec specifies scalar and `dtype: text` and test value is empty array/list -- ShapeError?
5) Spec specifies scalar and `dtype: text` and test value is empty string -- OK?

**Comment by oruebel** [comment]
> * Spec specifies `shape: None` and `dtype: float` and test value is empty array with type float -- ShapeError?

I think this would be a warning. ``shape: None`` does not seem to imply a minimum size, if it does then we would need a way to say that a dataset may be empty.

> * Spec specifies `shape: None` and `dtype: text` and test value is empty array with type float -- ShapeError and DtypeError?

This seems to be clearly a DtypeError first potentially combined with a warning that the dataset is empty.

> 
> * Spec specifies `shape: None` and `dtype: text` and test value is empty list -- ShapeError?
> 

Do you mean you don't have a dtype for the data in this case? I would issue warnings here.

> * Spec specifies scalar and `dtype: text` and test value is empty array/list -- ShapeError?

A scalar cannot be an array/list so a ShapeError seems appropriate. 

> 
> * Spec specifies scalar and `dtype: text` and test value is empty string -- OK?

Empty string is a valid string. I would issue a warning but not an error.

**Comment by rly** [comment]
Should the official pynwb validator issue these warnings for empty strings and empty datasets? Or should this instead be bundled into the best practices inspector? The validator currently does not issue any warnings.

**Comment by oruebel** [comment]
Since these are generic issues (i.e. issues that are not specific to the NWB schema), I think it would be fine to have such warnings be issued by the validator. It may be worthwhile to have this behavior configurable to enable/disable warnings.

**Comment by rly** [comment]
**Status update on this issue.**

The originally-reported `IndexError` is fixed (PR #377, merged June 2020). `get_type` now raises a private `EmptyArrayError` on empty input, which `AttributeValidator` and `DatasetValidator` catch and skip. The icephys "empty `DynamicTable` columns dataset" reproduction validates without crashing.

The five "what should the behavior be?" cases from the issue body, as of `dev` today, behave like this (verified with a minimal validator script):

| # | Spec | Value | Expected (per issue) | Current |
|---|------|-------|---------------------|---------|
| 1 | `shape: [None]`, `dtype: float` | empty `np.array([])` (float64) | OK | OK |
| 2 | `shape: [None]`, `dtype: text` | empty `np.array([])` (float64) | DtypeError | **silent, no error, no warning** |
| 3 | `shape: [None]`, `dtype: text` | empty `[]` | (open question) | silent |
| 4 | scalar, `dtype: text` | empty list / array | ShapeError | **silent** |
| 5 | scalar, `dtype: text` | `""` | OK | OK |

So the original crash is gone, but two real gaps remain.

1. **Empty typed array vs. wrong-dtype spec passes silently** (case 2). An empty `float64` array validated against a `text` spec emits no error and no warning. PRs #318 and #319 both proposed fixes here. #318's idea (return `data.dtype` for empty `np.ndarray` instead of raising `EmptyArrayError`) is the smaller, cleaner change and would now go in the `np.ndarray` branch of `get_type` around `validator.py:192`. It would need to be paired with a decision at the `EmptyArrayError` catch sites (`validator.py:407,431,549`) about whether to also warn when the type genuinely cannot be determined, e.g. for an empty Python list.

2. **Scalar spec accepts a list** (case 4). `dtype: text` with no shape and value `["a"]` or `[]` returns no error. Nothing in the current code raises on this. `check_shape` happily accepts any shape when `expected is None`. PR #319 introduced an `ExpectedScalarError` for this; it was never picked up. This needs a small change in `check_shape` (treat `expected is None` as "must be scalar") plus the new error class.

Both #318 and #319 are over five years old and don't apply cleanly to today's `validator.py`, so they should be closed as stale. The two fixes above are small enough to land as a fresh, narrower PR. Closing this issue should wait until those are in.

**Comment by rly** [comment]
**Correction to my previous comment.**

I tried to land the two "remaining gaps" and walked them back. Updated assessment, with the spec-language transition in mind.

### Case 4 (scalar spec accepting a list): correct per current spec, but the codebase is partway through a migration.

Per the current HDMF spec language, `shape: null` (or no `shape` key) means **scalar**, so by the strict reading, a `dtype: text` spec with no shape should reject `["a"]`. My intent was to enforce that.

In practice, several types in `hdmf-common-schema` and elsewhere have been written assuming "no shape = any shape allowed":
- `Data` in `common/base.yaml` has no shape and is used canonically for arrays (`Data('data1', [0, 1, 2, 3, 4])`).
- `VectorIndex.target` is a reference attribute with no shape; `get_data_shape` resolves it to the target dataset's `maxshape` like `(None, 2, 3)`.
- Many test specs declare typed datasets/attributes with no shape and pass arrays.

The plan is to migrate the spec language so `shape: null` means **any shape** and `shape: scalar` (new) means scalar — that hasn't been implemented. Until it is, enforcing strict scalar semantics in the validator generates false positives across these existing types.

So fixing case 4 cleanly is gated on the spec-language work. Once `shape: scalar` exists and the affected types have been updated (`Data` → some "any shape" form, `VectorIndex.target` → explicit scalar, etc.), the validator change is straightforward.

### Case 2 (empty typed array vs wrong-dtype spec): gated on a write-path bug.

I tried #318's approach: have `get_type` return `data.dtype` for empty `np.ndarray` instead of raising `EmptyArrayError`. It generates false positives on roundtrip because **HDMF writes an empty Python list as `float64` regardless of the spec's dtype**. Verified directly: `attr2: []` written against `dtype: int, shape: (None,)` lands as `float64` on disk. The validator's deliberate silence on empty-array dtypes (the `EmptyArrayError`-skip pattern) is a workaround for this, as the comment in `validator.py` ("do not validate dtype of empty array. HDMF does not yet set dtype when writing a list/tuple") has flagged since 2020.

Fixing case 2 requires fixing the write path first so empty data carries the spec's dtype to disk. Separate PR.

### Net status

- Original `IndexError` on empty arrays: fixed (#377).
- Case 4: blocked on the `shape: null` → any-shape / `shape: scalar` → scalar spec-language migration, plus updates to the types that currently rely on the lax interpretation.
- Case 2: blocked on a separate write-path dtype-loss bug.

Leaving this issue open as the umbrella, but the actionable next steps live in those two prerequisite work items.
